### PR TITLE
Feature/datasharing custom path [12051]

### PR DESCRIPTION
**Pull request by IkerLuengo**

### CODE REVIEW

File: src/cpp/rtps/DataSharing/DataSharingNotification.cpp [hunk 1]
```diff
@@ -50,101 +50,36 @@ std::shared_ptr<DataSharingNotification> DataSharingNotification::open_notificat
...
-        Segment::remove(segment_name_);
-
-        logError(HISTORY_DATASHARING_LISTENER, "Failed to create listener queue " << segment_name_
-                                                                                  << ": " << e.what());
-        return false;
-    }
-
-    owned_ = true;
-    return true;
+    return create_and_init_shared_segment_notification<fastdds::rtps::SharedMemSegment>(reader_guid,
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingNotification.cpp:57]: Could this sometimes be `SharedFileSegment` ?

File: src/cpp/rtps/DataSharing/DataSharingNotification.cpp [hunk 2]
```diff
@@ -50,101 +50,36 @@ std::shared_ptr<DataSharingNotification> DataSharingNotification::open_notificat
...
+                        shared_dir);
 }
 
 bool DataSharingNotification::open_and_init_notification(
         const GUID_t& reader_guid,
         const std::string& shared_dir)
 {
-    segment_id_ = reader_guid;
-    segment_name_ = generate_segment_name(shared_dir, reader_guid);
+    return open_and_init_shared_segment_notification<fastdds::rtps::SharedMemSegment>(reader_guid,
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/DataSharingNotification.cpp:65]: Could this sometimes be `SharedFileSegment` ?

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 could you update the result from trial? thanks in advance.

**Comment by Barry-Xu-2018** [comment]
I have tried this PR.  
Specified shared memory path can work correctly.   

After specify `shared_dir` in configuration file and execute, I can find   
```
$ ls /mnt/my_shm/   # Shared memory for data sharing is created in specified path
fast_datasharing_01.0f.eb.7d.04.18.1e.44.01.00.00.00_0.0.12.3 fast_datasharing_01.0f.eb.7d.10.18.f0.37.01.00.00.00_0.0.12.4

$ ls /dev/shm/   # Other control information is located at /dev/shm 
fastrtps_6e619c2c4a9ed965     fastrtps_a573b851cb1ef566_el  fastrtps_port7413     fastrtps_port7414_el  sem.fastrtps_port7412_mutex  sem.fastrtps_port7415_mutex
fastrtps_6e619c2c4a9ed965_el  fastrtps_port7412             fastrtps_port7413_el  fastrtps_port7415     sem.fastrtps_port7413_mutex
fastrtps_a573b851cb1ef566     fastrtps_port7412_el          fastrtps_port7414     fastrtps_port7415_el  sem.fastrtps_port7414_mutex
```

**Comment by IkerLuengo** [comment]
@Barry-Xu-2018 @fujitatomoya 

Just for clarification, the files that can still be found in `/dev/shm` are related to the SharedMemory transport, and not to the DataSharing. This is as designed, since we left the shared memory transport as it was, using the standard shared memory configuration.

File: src/cpp/rtps/DataSharing/DataSharingNotification.cpp (hunk 1 above)
Comment (IkerLuengo) [reply src/cpp/rtps/DataSharing/DataSharingNotification.cpp]: Absolutely. This was lost on 88cde53. Corrected on bdc5d2e

File: src/cpp/rtps/DataSharing/DataSharingNotification.cpp (hunk 2 above)
Comment (IkerLuengo) [reply src/cpp/rtps/DataSharing/DataSharingNotification.cpp]: Same as before

**Comment by fujitatomoya** [comment]
@IkerLuengo 

thanks for the comments, so we can actually ignore those files under `/dev/shm` for Data Sharing, is that correct. (just checking and reconfirmation.)

**Comment by IkerLuengo** [comment]
@fujitatomoya 
Yes, those appear because the shared memory transport is always on by default.
Files related to data sharing all have the pattern `fast_datasharing_.*`. The bunch of numbers following is the GUID of the endpoint. There should be one for each datawriter (for sharing the messages) and datareader (for notifications).

**Comment by IkerLuengo** [comment]
@richiprosima please test windows

**Comment by IkerLuengo** [comment]
Checked the failing tests. Errors are not related to the changes.

Windows job seems to be unstable and fails a lot due to `cmd.exe` failing during configuration of foonathan. However, we have complete results for each toolset on separate consecutive runs:

- http://jenkins.eprosima.com:8080/job/fastdds_windows/label=windows-secure,toolset=v141/1105/
- http://jenkins.eprosima.com:8080/job/fastdds_windows/label=windows-secure,toolset=v142/1104/

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**